Add some unit tests to the node and swarm cli code

vdemeester opened the pull request:
Start work on adding unit tests to our cli code in order to have to write less costly integration test. This is a little bit an experiment and adds a few things.

* Concept of *golden files* : A golden test is a fancy way of doing assert on a string or a directory tree, where the expected output is kept in a separate file or files -- the golden files. The basic idea is to use that for some cli outputs that would be hard to maintain otherwise. In that PR, if we want to update a golden file (because we changed on purpose the output of the test), we just have to run `go test -test.update` instead of `go test`.
* Builders for some struct (here `node` and `task`), to make test a little bit more readable but not having a hug struct in the middle of it.

```
go test --coverprofile=/home/vincent/go/src/github.com/docker/docker/cli/command/node/cover.out ./...
ok  	github.com/docker/docker/cli/command/node	0.013s	coverage: 97.2% of statements

coverage: 96.9% of statements
ok  	github.com/docker/docker/cli/command/swarm	0.024s
```

By having these kind of the tests, we reduce the number of integration tests required and thus should be able to lower a little bit the CI time (or at least use time better with more useful integration tests) — it all started with the will to remove `TestSwarmNodeListHostname` from integration test (this test had nothing to do as an integration one).

One wish would be to switch to 1.7 at some point to use subtest with these table-driver test, that would be cool 👼 :sunglasses:

/cc @icecrime @dnephin @cpuguy83 @thaJeztah 

🐸

Signed-off-by: Vincent Demeester <vincent@sbr.pm>

dnephin commented on cli/command/node/builders_test.go:
```diff
@@ -0,0 +1,149 @@
+package node
```
This node builder would probably be useful across a few packages. What do you think about moving it to a new package and exporting it?

vdemeester replied:
Yeah, I though about that, but didn't want to get too much ahead and refactor as I go in other PR. I don't want to over-design if it's not needed.

vdemeester replied:
even if it's needed, just wanted to go with the flow 😛

dnephin commented on cli/command/node/client_test.go:
```diff
@@ -0,0 +1,85 @@
+	command.DockerCli
+	client client.APIClient
+	out    io.Writer
+}
```
This could definitely be used by unit tests in other packages.

What do you think about using https://github.com/vektra/mockery to generate something like this instead of writing them manually?

vdemeester replied:
Same as above, I definitely think it could be useful for othe tests. It felt simpler and easier to use a fake than use another dependency. If it gets too complicated I would be ok to use it and/or generate it 👼 (I tend to prefer fake to mock when it's simpler to write — but yeah should take a closer look at `mockery` in the near future).

aaronlehmann commented:
I like what I see here. We've had very little coverage of CLI behavior, so it's great to get more. This could help us avoid regressions.

> By having these kind of the tests, we reduce the number of integration tests required and thus should be able to lower a little bit the CI time

I see that you removed `TestSwarmNodeListHostname`, and I agree with that, as it's rather silly for an integration test. But I'd caution against taking this too far. We still need integration tests that confirm the CLI works correctly against a real swarm. I would argue we need a lot more than we currently have.

dnephin commented on cli/command/node/ps_test.go:
```diff
@@ -0,0 +1,142 @@
+				},
+				expectedError: "error returning the task list",
+			},
+		*/
```
Commented out code.

vdemeester replied:
Yeah, related to the *should it be an error* question, I should remove it yeah 👍

dnephin commented on pkg/testutil/golden/golden.go:
```diff
@@ -0,0 +1,28 @@
+	"testing"
+)
+
+var update = flag.Bool("test.update", false, "update golden file")
```
I think it's kind of bad form to register CLI flags from arbitrary non-main packages. 

I really like the idea of being able to update these files easily, but I wonder if we could do this with a separate tool

vdemeester replied:
yeah same for me, but as it's a testing package I thought it was a trade-off I was willing to take.

vdemeester replied:
and `go test -test.update` is cool 😛 😎

dnephin reviewed (commented):
I really like this.

dnephin commented:
> I'd caution against taking this too far. We still need integration tests that confirm the CLI works correctly against a real swarm. I would argue we need a lot more than we currently have.

We should have some clear guidelines about when to write an integration test. An integration test for each important "feature" and important error cases is appropriate. Adding one for every CLI flag is overkill.

vdemeester commented:
> I see that you removed `TestSwarmNodeListHostname`, and I agree with that, as it's rather silly for an integration test. But I'd caution against taking this too far. We still need integration tests that confirm the CLI works correctly against a real swarm. I would argue we need a lot more than we currently have.

@aaronlehmann yes definitely. I just want to remove the one that can be done client side, and make sure the one we have are at the right place and covers it ok.

@dnephin yes 👼

vdemeester commented:
@dnephin started some work on the `cli/command/swarm` package to see what needs to be refactored https://github.com/docker/docker/compare/master...vdemeester:swarm-cli-unit-tests

vdemeester commented:
So… *golden files* and windows are not friends.. I might have to handle that in the `golden` package somehow…

vdemeester committed:
Add some unit tests to the node and swarm cli code

Start work on adding unit tests to our cli code in order to have to
write less costly integration test.

Signed-off-by: Vincent Demeester <vincent@sbr.pm>

LK4D4 commented:
need rebase

vdemeester commented:
Rebase 👼

thaJeztah commented:
ping @aaronlehmann @dnephin PTAL

dnephin commented on cli/internal/test/builder/node.go:
```diff
@@ -0,0 +1,117 @@
+	return func(node *swarm.Node) {
+		node.ID = id
+	}
+}
```
What do you think about making these return the `node` instead of just mutating it? 

They could take a non-pointer type as well so that you can pass in intermediate instances if you want to build multiple instances.

aaronlehmann replied:
> What do you think about making these return the node instead of just mutating it?

That gets a bit complicated when mutating things beneath pointer indirections or slices/maps. For example, `NodeLabels` would have to copy the items in the `Labels` map to a new map before mutating it.

For the native gRPC types we code-gen `Copy` methods that do deep copies. They would be useful for something like this. Unfortunately the engine types don't have this.

aaronlehmann commented on cli/internal/test/builder/node.go:
```diff
@@ -0,0 +1,117 @@
+package builder
```
You might consider finding a different name so this is not confused with the builder that implements `docker build`, but I don't have any ideas to offer.

aaronlehmann replied:
Maybe `constructor`?

aaronlehmann commented on cli/command/node/inspect_test.go:
```diff
@@ -0,0 +1,120 @@
+	"github.com/docker/docker/pkg/testutil/golden"
+)
+
+func TestNodeInspectShouldReturnAnErrorIfAPIfail(t *testing.T) {
```
`Fail` should be uppercase.

Perhaps shorten the name to `TestNodeInspectErrorOnAPIFailure`

aaronlehmann commented on cli/command/node/list_test.go:
```diff
@@ -0,0 +1,99 @@
+	"github.com/docker/docker/pkg/testutil/assert"
+)
+
+func TestNodeListShouldReturnAnErrorIfAPIfail(t *testing.T) {
```
See https://github.com/docker/docker/pull/29143/files#r94501063

aaronlehmann commented on cli/command/node/remove_test.go:
```diff
@@ -0,0 +1,45 @@
+	}
+}
+
+func TestNodeRemoveMultipleNode(t *testing.T) {
```
`TestNodeRemoveMultiple` or `TestNodeRemoveMultipleNodes`

aaronlehmann commented on cli/command/swarm/leave_test.go:
```diff
@@ -0,0 +1,52 @@
+		expectedError  string
+	}{
+		{
+			name:          "too-much-args",
```
`too-many-args`

aaronlehmann commented on cli/command/swarm/unlock_key_test.go:
```diff
@@ -0,0 +1,175 @@
+		expectedError         string
+	}{
+		{
+			name:          "too-much-args",
```
`too-many-args`

aaronlehmann commented on cli/command/swarm/unlock_test.go:
```diff
@@ -0,0 +1,101 @@
+		expectedError   string
+	}{
+		{
+			name:          "too-much-args",
```
`too-many-args`

aaronlehmann commented on cli/command/swarm/unlock_test.go:
```diff
@@ -0,0 +1,101 @@
+	"github.com/docker/docker/pkg/testutil/assert"
+)
+
+func TestSwarmUnlockShouldReturnAnErrorIfAPIfail(t *testing.T) {
```
Same comment as above about the naming of this test.

aaronlehmann commented on cli/command/swarm/update_test.go:
```diff
@@ -0,0 +1,182 @@
+		expectedError         string
+	}{
+		{
+			name:          "too-much-args",
```
`too-many-args`

aaronlehmann commented on cli/command/swarm/join_test.go:
```diff
@@ -0,0 +1,102 @@
+			expectedError: "requires exactly 1 argument",
+		},
+		{
+			name:          "too-much-args",
```
`too-many-args`

aaronlehmann commented on cli/command/swarm/join_test.go:
```diff
@@ -0,0 +1,102 @@
+	"github.com/docker/docker/pkg/testutil/assert"
+)
+
+func TestSwarmJoinShouldReturnAnErrorIfAPIfail(t *testing.T) {
```
See other comments about naming this test.

aaronlehmann commented on cli/command/swarm/join_token_test.go:
```diff
@@ -0,0 +1,215 @@
+			expectedError: "requires exactly 1 argument",
+		},
+		{
+			name:          "too-much-args",
```
`too-many-args`

aaronlehmann commented on cli/command/swarm/leave_test.go:
```diff
@@ -0,0 +1,52 @@
+	"github.com/docker/docker/pkg/testutil/assert"
+)
+
+func TestSwarmLeaveShouldReturnAnErrorIfAPIfail(t *testing.T) {
```
`Fail` should be uppercase, consider shortening the name.

aaronlehmann commented on cli/command/swarm/init_test.go:
```diff
@@ -0,0 +1,129 @@
+	"github.com/docker/docker/pkg/testutil/golden"
+)
+
+func TestSwarmInitShouldReturnAnErrorIfAPIfail(t *testing.T) {
```
`Fail` should be uppercase, consider shortening the name.

aaronlehmann commented:
Very happy with this overall. Have a few comments about wording and naming.

vdemeester commented on pkg/testutil/assert/assert.go:
```diff
@@ -25,6 +26,25 @@ func Equal(t TestingT, actual, expected interface{}) {
+// transform function. It fails the test if these two transformed string are not equal.
+// For example `EqualTransformedString(t, RemoveSpace, "foo\n", "foo")` wouldn't fail the test as
+// spaces (and thus '\n') are removed before comparing the string.
+func EqualTransformedString(t TestingT, transformFun func(rune) rune, actual, expected string) {
```
I would be better 👍

thaJeztah commented:
ping @aaronlehmann @dnephin PR was updated, PTAL

dnephin reviewed (approved):
LGTM

aaronlehmann commented on cli/internal/test/builders/node.go:
```diff
@@ -0,0 +1,117 @@
+package builder
```
Should the package name be `builders`?

vdemeester replied:
@aaronlehmann ah right, I knew I forgot something 😅

vdemeester replied:
(and updated)

aaronlehmann reviewed (approved):
LGTM

thaJeztah commented:
all 💚
